From 07c69d869f7235ed1a227ff12880ac8696da14f5 Mon Sep 17 00:00:00 2001
From: Liu Ying <victor.liu@nxp.com>
Date: Mon, 11 Jan 2016 10:51:21 +0800
Subject: [PATCH 2/2] MLK-12142 video: fbdev: mxc_ipuv3_fb: Set non-zero blank
 to FB_BLANK_POWERDOWN

commit  cbcd36d158a9155f4b023b4a45129f2ae5cf0354 from
https://source.codeaurora.org/external/imx/linux-imx.git

We have implemented only two states for blank which are FB_BLANK_UNBLANK and
FB_BLANK_POWERDOWN.  Let's map all the other blank states(FB_BLANK_NORMAL,
FB_BLANK_VSYNC_SUSPEND and FB_BLANK_HSYNC_SUSPEND) to FB_BLANK_POWERDOWN,
since all of them make the screen be blanked.  This may avoid blanking a blanked
screen by setting the state to different non-zero values.

Signed-off-by: Liu Ying <victor.liu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/video/fbdev/mxc/mxc_ipuv3_fb.c | 21 +++++++++------------
 1 file changed, 9 insertions(+), 12 deletions(-)

diff --git a/drivers/video/fbdev/mxc/mxc_ipuv3_fb.c b/drivers/video/fbdev/mxc/mxc_ipuv3_fb.c
index 1de38962415e..d125d493e83c 100644
--- a/drivers/video/fbdev/mxc/mxc_ipuv3_fb.c
+++ b/drivers/video/fbdev/mxc/mxc_ipuv3_fb.c
@@ -1,5 +1,5 @@
 /*
- * Copyright 2004-2015 Freescale Semiconductor, Inc. All Rights Reserved.
+ * Copyright 2004-2016 Freescale Semiconductor, Inc. All Rights Reserved.
  */
 
 /*
@@ -2316,16 +2316,19 @@ static int mxcfb_blank(int blank, struct fb_info *info)
 
 	dev_dbg(info->device, "blank = %d\n", blank);
 
+	if (blank)
+		blank = FB_BLANK_POWERDOWN;
+
 	if (mxc_fbi->cur_blank == blank)
 		return 0;
 
 	mxc_fbi->next_blank = blank;
 
-	switch (blank) {
-	case FB_BLANK_POWERDOWN:
-	case FB_BLANK_VSYNC_SUSPEND:
-	case FB_BLANK_HSYNC_SUSPEND:
-	case FB_BLANK_NORMAL:
+	if (blank == FB_BLANK_UNBLANK) {
+		info->var.activate = (info->var.activate & ~FB_ACTIVATE_MASK) |
+				FB_ACTIVATE_NOW | FB_ACTIVATE_FORCE;
+		ret = fb_set_var(info, &info->var);
+	} else {
 		if (mxc_fbi->dispdrv && mxc_fbi->dispdrv->drv->disable)
 			mxc_fbi->dispdrv->drv->disable(mxc_fbi->dispdrv, info);
 		ipu_disable_channel(mxc_fbi->ipu, mxc_fbi->ipu_ch, true);
@@ -2337,12 +2340,6 @@ static int mxcfb_blank(int blank, struct fb_info *info)
 			ipu_pre_disable(mxc_fbi->pre_num);
 			ipu_pre_free(&mxc_fbi->pre_num);
 		}
-		break;
-	case FB_BLANK_UNBLANK:
-		info->var.activate = (info->var.activate & ~FB_ACTIVATE_MASK) |
-				FB_ACTIVATE_NOW | FB_ACTIVATE_FORCE;
-		ret = fb_set_var(info, &info->var);
-		break;
 	}
 	if (!ret)
 		mxc_fbi->cur_blank = blank;
-- 
2.17.1

